internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
 			"rejected bad transaction",
-			"priority", wtx.priority,
+			"priority", wtx.Priority(),
 			"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
-			"peer_id", txInfo.SenderNodeID,
+			"peer_id", wtx.peers,
 			"code", checkTxRes.CheckTx.Code,
 			"post_check_err", err,
 		)
 
 		txmp.metrics.FailedTxs.Add(1)
 
+		// Remove the invalid transaction from the cache, unless the operator has
+		// instructed us to keep invalid transactions.
 		if !txmp.config.KeepInvalidTxsInCache {
 			txmp.cache.Remove(wtx.tx)
 		}
+
+		// If there was a post-check error, record its text in the result for
+		// debugging purposes.
 		if err != nil {
 			checkTxRes.CheckTx.MempoolError = err.Error()
 		}
 		return
 	}
 
-	sender := checkTxRes.CheckTx.Sender
 	priority := checkTxRes.CheckTx.Priority
+	sender := checkTxRes.CheckTx.Sender
 
-	if len(sender) > 0 {
-		if wtx := txmp.txStore.GetTxBySender(sender); wtx != nil {
+	// Disallow multiple concurrent transactions from the same sender assigned
+	// by the ABCI application. As a special case, an empty sender is not
+	// restricted.
+	if sender != "" {
+		elt, ok := txmp.txBySender[sender]
+		if ok {
+			w := elt.Value.(*WrappedTx)
 			txmp.logger.Error(
-				"rejected incoming good transaction; tx already exists for sender",
-				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+				"rejected valid incoming transaction; tx already exists for sender",
+				"tx", fmt.Sprintf("%X", w.tx.Hash()),
 				"sender", sender,
 			)
-			txmp.metrics.RejectedTxs.Add(1)
 			return
 		}
 	}
 
+	// At this point the application has ruled the transaction valid, but the
+	// mempool might be full. If so, find the lowest-priority items with lower
+	// priority than the application assigned to this new one, and evict as many
+	// of them as necessary to make room for tx. If no such items exist, we
+	// discard tx.
+
 	if err := txmp.canAddTx(wtx); err != nil {
-		evictTxs := txmp.priorityIndex.GetEvictableTxs(
-			priority,
-			int64(wtx.Size()),
-			txmp.SizeBytes(),
-			txmp.config.MaxTxsBytes,
-		)
-		if len(evictTxs) == 0 {
-			// No room for the new incoming transaction so we just remove it from
-			// the cache.
+		var victims []*clist.CElement // eligible transactions for eviction
+		var victimBytes int64         // total size of victims
+		for cur := txmp.txs.Front(); cur != nil; cur = cur.Next() {
+			cw := cur.Value.(*WrappedTx)
+			if cw.priority < priority {
+				victims = append(victims, cur)
+			}
+		}
+
+		// If there are no suitable eviction candidates, or the total size of
+		// those candidates is not enough to make room for the new transaction,
+		// drop the new one.
+		if len(victims) == 0 || victimBytes < wtx.Size() {
 			txmp.cache.Remove(wtx.tx)
 			txmp.logger.Error(
-				"rejected incoming good transaction; mempool full",
+				"rejected valid incoming transaction; mempool is full",
 				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
 				"err", err.Error(),
 			)
 			txmp.metrics.RejectedTxs.Add(1)
 			return
 		}
```

Comment:
Can we add an error to `mempool_error` indicating that this transaction was dropped because the mempool is full and it is the lowest priority